File: 02.SvelteElectron/apps/ui/src/shared/utils/postseasonEngine.ts
import type { PostseasonSeries, ScheduleEntry, Standing } from "../types/season";
import { seedOf } from "./seedOf";

// ── 순수 TS 헬퍼 (계산만, 상태 없음) ────────────────────────────
export function winsNeeded(bestOf: 1 | 3 | 5 | 7): number {
  return Math.ceil(bestOf / 2);
}

export function nextGameNum(series: PostseasonSeries): number {
  return series.homeWins + series.awayWins + 1;
}

export function getActiveSeries(bracket: PostseasonSeries[]): PostseasonSeries | null {
  return bracket.find((s) => !s.winner && s.homeTeamId !== "" && s.awayTeamId !== "") ?? null;
}

// ── Rust IPC 래퍼 ────────────────────────────────────────────────
export async function buildKblBracket(standings: Standing[]): Promise<PostseasonSeries[]> {
  const raw = await window.projectB!.postseasonBuildKbl(
    JSON.stringify({ standings })
  );
  return JSON.parse(raw);
}

export async function buildAblBracket(
  eastStandings: Standing[],
  westStandings: Standing[],
): Promise<PostseasonSeries[]> {
  const raw = await window.projectB!.postseasonBuildAbl(
    JSON.stringify({ eastStandings, westStandings })
  );
  return JSON.parse(raw);
}

// 고교 buildHsBracket · 대학 buildUnivBracket은 Phase 5-5a에서 폐기했다.
// 시즌 결산이 패왕기(고교 11월)·왕중왕전(대학 5월)로 옮겨갔다 — top4 준결승/결승을
// 남겨두면 결승이 두 번 열린다. Rust build_hs_bracket / build_univ_bracket도 제거됨.

/**
 * 프로 2군 축약 포스트시즌 (Phase 5-7).
 *
 * 상위 4팀 단판 사다리(3위vs4위 → 승자vs2위 → 승자vs1위).
 * 독립 사다리와 모양은 같지만 **결승도 단판**이다(독립 챔결은 3전2승).
 */
export async function buildFarmBracket(standings: Standing[]): Promise<PostseasonSeries[]> {
  const raw = await window.projectB!.engine(
    "buildFarmBracketNative", JSON.stringify({ standings }),
  );
  return JSON.parse(raw);
}

/**
 * 독립 4차 Stage 사다리 (Phase 5-6).
 *
 * 준PO(3위 vs 4위, 단판) → PO(2위 vs 준PO승자, 단판) → 챔피언결정전(1위 vs PO승자, 3전2승).
 * 구 buildIndBracket은 "1위 vs 2위 단판" 하나뿐이라 4팀 사다리를 표현하지 못했다.
 */
export async function buildIndLadder(standings: Standing[]): Promise<PostseasonSeries[]> {
  const raw = await window.projectB!.engine(
    "buildIndLadderNative", JSON.stringify({ standings }),
  );
  return JSON.parse(raw);
}

export async function buildJblBracket(standings: Standing[]): Promise<PostseasonSeries[]> {
  const raw = await window.projectB!.postseasonBuildJbl(
    JSON.stringify({ standings })
  );
  return JSON.parse(raw);
}

export async function applyGameToSeries(
  series: PostseasonSeries,
  winnerId: string,
): Promise<PostseasonSeries> {
  const raw = await window.projectB!.postseasonApplyGame(
    JSON.stringify({ series, winnerId })
  );
  return JSON.parse(raw);
}

export async function fillNextSeries(
  bracket: PostseasonSeries[],
  completed: PostseasonSeries,
): Promise<PostseasonSeries[]> {
  const raw = await window.projectB!.postseasonFillNext(
    JSON.stringify({ bracket, completed })
  );
  return JSON.parse(raw);
}

/**
 * 비주인공 시리즈 씨앗.
 *
 * 🔴 **씨앗을 안 넘기면 Rust 가 `thread_rng` 로 떨어진다** — 같은 세이브를
 *   두 번 돌리면 **다른 팀이 우승한다**(2026-09-07 실측: 독립리그 준PO 승자가
 *   판마다 갈렸다). 엔진에는 `seed` 자리가 처음부터 있었는데 **부르는 쪽이
 *   안 채우고 있었다.**
 *
 * ⚠ **판마다 달라야 한다.** 한 포스트시즌에서 이 함수는 여러 번 불린다
 *   (브라켓 생성 직후 · 주인공 탈락 뒤 · 주인공 시리즈가 끝날 때마다).
 *   같은 씨앗을 그대로 재사용하면 LCG 가 매번 처음부터 같은 수열을 내서
 *   **매 라운드 홈이 이기는** 편향이 생긴다. 그래서 **이미 승자가 정해진
 *   시리즈 수**를 섞는다 — 브라켓이 나아갈 때마다 달라지는 값이다.
 */
export function postseasonSeed(
  worldSeed: number,
  seasonYear: number,
  leagueId: string,
  bracket: PostseasonSeries[],
): number {
  return seedOf(worldSeed, seasonYear, leagueId, "postseason",
    bracket.filter((s) => s.winner).length);
}

export async function resolveNonProtagonistSeries(
  bracket: PostseasonSeries[],
  protagonistTeamId: string,
  seed = 0,
): Promise<PostseasonSeries[]> {
  const raw = await window.projectB!.postseasonResolveNpc(
    JSON.stringify({ bracket, protagonistTeamId, seed })
  );
  return JSON.parse(raw);
}

export async function makeSeriesGame(
  series: PostseasonSeries,
  gameNum: number,
  baseWeek: number,
  protagonistTeamId: string,
  seasonYear: number,
): Promise<ScheduleEntry> {
  const raw = await window.projectB!.postseasonMakeGame(
    JSON.stringify({ series, gameNum, baseWeek, protagonistTeamId, seasonYear })
  );
  return JSON.parse(raw);
}

export async function shuffleAblConferences(
  allTeams: string[],
): Promise<{ east: string[]; west: string[] }> {
  const raw = await window.projectB!.postseasonShuffleAbl(
    JSON.stringify({ allTeams })
  );
  return JSON.parse(raw);
}

File: 02.SvelteElectron/packages/engine-native/src/postseason_engine.rs
use serde::{Deserialize, Serialize};
use rand::Rng;
use crate::schedule_engine::{ScheduleEntry, to_game_date};

// ── Types ─────────────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Standing {
    pub team_id: String,
    pub wins: u32,
    pub losses: u32,
    pub draws: u32,
    pub win_pct: f64,
    pub runs_for: u32,
    pub runs_against: u32,
    pub streak: String,
    pub last10: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct PostseasonSeries {
    pub id: String,
    pub league_id: String,
    pub round: String,
    pub home_team_id: String,
    pub away_team_id: String,
    pub best_of: u32,
    pub home_wins: u32,
    pub away_wins: u32,
    pub winner: Option<String>,
    pub home_from: Option<String>,
    pub away_from: Option<String>,
    pub next_series_id: Option<String>,
    pub next_series_slot: Option<String>,
}

fn wins_needed(best_of: u32) -> u32 {
    (best_of + 1) / 2
}

fn sort_teams(standings: &[Standing]) -> Vec<String> {
    let mut sorted = standings.to_vec();
    sorted.sort_by(|a, b| {
        b.win_pct.partial_cmp(&a.win_pct)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then(b.wins.cmp(&a.wins))
    });
    sorted.into_iter().map(|s| s.team_id).collect()
}

// ── Bracket builders ──────────────────────────────────────────

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BuildBracketParams {
    pub standings: Vec<Standing>,
}

pub fn build_kbl_bracket(p: BuildBracketParams) -> Vec<PostseasonSeries> {
    let t = sort_teams(&p.standings);
    if t.len() < 6 { return vec![]; }
    vec![
        PostseasonSeries { id: "KBL_WC".into(),   league_id: "LEAGUE_KBL".into(), round: "와일드카드".into(),   home_team_id: t[4].clone(), away_team_id: t[5].clone(), best_of: 1, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: None,                    next_series_id: Some("KBL_PREP".into()), next_series_slot: Some("away".into()) },
        PostseasonSeries { id: "KBL_PREP".into(), league_id: "LEAGUE_KBL".into(), round: "준플레이오프".into(), home_team_id: t[3].clone(), away_team_id: "".into(),     best_of: 3, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: Some("KBL_WC".into()),   next_series_id: Some("KBL_PO".into()),   next_series_slot: Some("away".into()) },
        PostseasonSeries { id: "KBL_PO".into(),   league_id: "LEAGUE_KBL".into(), round: "플레이오프".into(),   home_team_id: t[1].clone(), away_team_id: "".into(),     best_of: 5, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: Some("KBL_PREP".into()), next_series_id: Some("KBL_KS".into()),   next_series_slot: Some("away".into()) },
        PostseasonSeries { id: "KBL_KS".into(),   league_id: "LEAGUE_KBL".into(), round: "한국시리즈".into(),   home_team_id: t[0].clone(), away_team_id: "".into(),     best_of: 7, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: Some("KBL_PO".into()),   next_series_id: None,                    next_series_slot: None },
    ]
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BuildAblBracketParams {
    pub east_standings: Vec<Standing>,
    pub west_standings: Vec<Standing>,
}

pub fn build_abl_bracket(p: BuildAblBracketParams) -> Vec<PostseasonSeries> {
    let e = sort_teams(&p.east_standings);
    let w = sort_teams(&p.west_standings);
    if e.len() < 4 || w.len() < 4 { return vec![]; }
    vec![
        PostseasonSeries { id: "ABL_EWC".into(), league_id: "LEAGUE_ABL".into(), round: "이스트 와일드카드".into(),    home_team_id: e[2].clone(), away_team_id: e[3].clone(), best_of: 3, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: None,                    next_series_id: Some("ABL_EDS".into()), next_series_slot: Some("away".into()) },
        PostseasonSeries { id: "ABL_WWC".into(), league_id: "LEAGUE_ABL".into(), round: "웨스트 와일드카드".into(),    home_team_id: w[2].clone(), away_team_id: w[3].clone(), best_of: 3, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: None,                    next_series_id: Some("ABL_WDS".into()), next_series_slot: Some("away".into()) },
        PostseasonSeries { id: "ABL_EDS".into(), league_id: "LEAGUE_ABL".into(), round: "이스트 디비전시리즈".into(), home_team_id: e[0].clone(), away_team_id: "".into(),     best_of: 5, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: Some("ABL_EWC".into()), next_series_id: Some("ABL_CS".into()),  next_series_slot: Some("home".into()) },
        PostseasonSeries { id: "ABL_WDS".into(), league_id: "LEAGUE_ABL".into(), round: "웨스트 디비전시리즈".into(), home_team_id: w[0].clone(), away_team_id: "".into(),     best_of: 5, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: Some("ABL_WWC".into()), next_series_id: Some("ABL_CS".into()),  next_series_slot: Some("away".into()) },
        PostseasonSeries { id: "ABL_CS".into(),  league_id: "LEAGUE_ABL".into(), round: "챔피언십시리즈".into(),      home_team_id: "".into(),    away_team_id: "".into(),     best_of: 7, home_wins: 0, away_wins: 0, winner: None, home_from: Some("ABL_EDS".into()), away_from: Some("ABL_WDS".into()), next_series_id: None, next_series_slot: None },
    ]
}

pub fn build_univ_bracket(p: BuildBracketParams) -> Vec<PostseasonSeries> {
    let t = sort_teams(&p.standings);
    if t.len() < 4 { return vec![]; }
    vec![
        PostseasonSeries { id: "UNIV_SEMI1".into(), league_id: "LEAGUE_UNIVERSITY".into(), round: "준결승".into(), home_team_id: t[0].clone(), away_team_id: t[3].clone(), best_of: 1, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: None, next_series_id: Some("UNIV_FINAL".into()), next_series_slot: Some("home".into()) },
        PostseasonSeries { id: "UNIV_SEMI2".into(), league_id: "LEAGUE_UNIVERSITY".into(), round: "준결승".into(), home_team_id: t[1].clone(), away_team_id: t[2].clone(), best_of: 1, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: None, next_series_id: Some("UNIV_FINAL".into()), next_series_slot: Some("away".into()) },
        PostseasonSeries { id: "UNIV_FINAL".into(), league_id: "LEAGUE_UNIVERSITY".into(), round: "결승".into(),   home_team_id: "".into(),    away_team_id: "".into(),    best_of: 1, home_wins: 0, away_wins: 0, winner: None, home_from: Some("UNIV_SEMI1".into()), away_from: Some("UNIV_SEMI2".into()), next_series_id: None, next_series_slot: None },
    ]
}

pub fn build_jbl_bracket(p: BuildBracketParams) -> Vec<PostseasonSeries> {
    let cl: Vec<Standing> = p.standings.iter().filter(|s| s.team_id.contains("_CL_")).cloned().collect();
    let pl: Vec<Standing> = p.standings.iter().filter(|s| s.team_id.contains("_PL_")).cloned().collect();
    if cl.len() < 3 || pl.len() < 3 { return vec![]; }
    let cl = sort_teams(&cl);
    let pl = sort_teams(&pl);
    vec![
        PostseasonSeries { id: "JBL_CL_CS1".into(), league_id: "LEAGUE_JBL".into(), round: "CL 클라이맥스 1스테이지".into(), home_team_id: cl[1].clone(), away_team_id: cl[2].clone(), best_of: 3, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: None,                       next_series_id: Some("JBL_CL_CF".into()), next_series_slot: Some("away".into()) },
        PostseasonSeries { id: "JBL_PL_CS1".into(), league_id: "LEAGUE_JBL".into(), round: "PL 클라이맥스 1스테이지".into(), home_team_id: pl[1].clone(), away_team_id: pl[2].clone(), best_of: 3, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: None,                       next_series_id: Some("JBL_PL_CF".into()),  next_series_slot: Some("away".into()) },
        PostseasonSeries { id: "JBL_CL_CF".into(),  league_id: "LEAGUE_JBL".into(), round: "CL 클라이맥스 파이널".into(),    home_team_id: cl[0].clone(), away_team_id: "".into(),     best_of: 5, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: Some("JBL_CL_CS1".into()), next_series_id: Some("JBL_JS".into()),    next_series_slot: Some("home".into()) },
        PostseasonSeries { id: "JBL_PL_CF".into(),  league_id: "LEAGUE_JBL".into(), round: "PL 클라이맥스 파이널".into(),    home_team_id: pl[0].clone(), away_team_id: "".into(),     best_of: 5, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: Some("JBL_PL_CS1".into()), next_series_id: Some("JBL_JS".into()),    next_series_slot: Some("away".into()) },
        PostseasonSeries { id: "JBL_JS".into(),     league_id: "LEAGUE_JBL".into(), round: "일본시리즈".into(),              home_team_id: "".into(),     away_team_id: "".into(),     best_of: 7, home_wins: 0, away_wins: 0, winner: None, home_from: Some("JBL_CL_CF".into()), away_from: Some("JBL_PL_CF".into()), next_series_id: None, next_series_slot: None },
    ]
}

pub fn build_ind_bracket(p: BuildBracketParams) -> Vec<PostseasonSeries> {
    let t = sort_teams(&p.standings);
    if t.len() < 2 { return vec![]; }
    vec![
        PostseasonSeries { id: "IND_FINAL".into(), league_id: "LEAGUE_INDEPENDENT".into(), round: "결승".into(), home_team_id: t[0].clone(), away_team_id: t[1].clone(), best_of: 1, home_wins: 0, away_wins: 0, winner: None, home_from: None, away_from: None, next_series_id: None, next_series_slot: None },
    ]
}

// ── Series operations ─────────────────────────────────────────

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyGameToSeriesParams {
    pub series: PostseasonSeries,
    pub winner_id: String,
}

pub fn apply_game_to_series(p: ApplyGameToSeriesParams) -> PostseasonSeries {
    let s = p.series;
    let hw = if p.winner_id == s.home_team_id { s.home_wins + 1 } else { s.home_wins };
    let aw = if p.winner_id == s.away_team_id { s.away_wins + 1 } else { s.away_wins };
    let needed = wins_needed(s.best_of);
    let winner = if hw >= needed { Some(s.home_team_id.clone()) }
                 else if aw >= needed { Some(s.away_team_id.clone()) }
                 else { None };
    PostseasonSeries { home_wins: hw, away_wins: aw, winner, ..s }
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FillNextSeriesParams {
    pub bracket: Vec<PostseasonSeries>,
    pub completed: PostseasonSeries,
}

pub fn fill_next_series(p: FillNextSeriesParams) -> Vec<PostseasonSeries> {
    let c = &p.completed;
    if c.winner.is_none() || c.next_series_id.is_none() { return p.bracket; }
    let next_id = c.next_series_id.as_deref().unwrap();
    let winner = c.winner.as_deref().unwrap();
    let slot = c.next_series_slot.as_deref();
    p.bracket.into_iter().map(|s| {
        if s.id != next_id { return s; }
        let home = if slot == Some("home") { winner.to_string() } else { s.home_team_id.clone() };
        let away = if slot == Some("away") { winner.to_string() } else { s.away_team_id.clone() };
        PostseasonSeries { home_team_id: home, away_team_id: away, ..s }
    }).collect()
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResolveNpcSeriesParams {
    pub bracket: Vec<PostseasonSeries>,
    pub protagonist_team_id: String,
}

pub fn resolve_non_protagonist_series(p: ResolveNpcSeriesParams) -> Vec<PostseasonSeries> {
    let mut cur = p.bracket;
    let pt = &p.protagonist_team_id;
    let mut rng = rand::thread_rng();

    for _ in 0..20 {
        let mut changed = false;
        for i in 0..cur.len() {
            if cur[i].winner.is_some() { continue; }
            if cur[i].home_team_id.is_empty() || cur[i].away_team_id.is_empty() { continue; }
            if &cur[i].home_team_id == pt || &cur[i].away_team_id == pt { continue; }

            let needed = wins_needed(cur[i].best_of);
            let mut hw = 0u32;
            let mut aw = 0u32;
            while hw < needed && aw < needed {
                if rng.gen::<f64>() < 0.5 { hw += 1; } else { aw += 1; }
            }
            let winner = if hw >= needed { Some(cur[i].home_team_id.clone()) } else { Some(cur[i].away_team_id.clone()) };
            let updated = PostseasonSeries { home_wins: hw, away_wins: aw, winner, ..cur[i].clone() };
            let completed = updated.clone();
            cur[i] = updated;
            cur = fill_next_series(FillNextSeriesParams { bracket: cur, completed });
            changed = true;
        }
        if !changed { break; }
    }
    cur
}

// ── Make series game ──────────────────────────────────────────

const SERIES_DAY_OFFSETS: &[u32] = &[5, 6, 8, 9, 10, 12, 13];

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MakeSeriesGameParams {
    pub series: PostseasonSeries,
    pub game_num: u32,
    pub base_week: u32,
    pub protagonist_team_id: String,
    pub season_year: u32,
}

pub fn make_series_game(p: MakeSeriesGameParams) -> ScheduleEntry {
    let off_idx = ((p.game_num - 1) as usize).min(SERIES_DAY_OFFSETS.len() - 1);
    let offset = SERIES_DAY_OFFSETS[off_idx];
    let week = if offset >= 7 { p.base_week + 1 } else { p.base_week };
    let s = &p.series;
    ScheduleEntry {
        id: format!("{}_G{}", s.id, p.game_num),
        week,
        game_date: to_game_date(p.season_year, p.base_week, offset),
        league_id: Some(s.league_id.clone()),
        home_team_id: s.home_team_id.clone(),
        away_team_id: s.away_team_id.clone(),
        is_protagonist_game: s.home_team_id == p.protagonist_team_id || s.away_team_id == p.protagonist_team_id,
        phase: "postseason".to_string(),
    }
}

// ── Shuffle ───────────────────────────────────────────────────

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AblConferences {
    pub east: Vec<String>,
    pub west: Vec<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ShuffleAblConferencesParams {
    pub all_teams: Vec<String>,
}

pub fn shuffle_abl_conferences(p: ShuffleAblConferencesParams) -> AblConferences {
    let mut arr = p.all_teams;
    let mut rng = rand::thread_rng();
    for i in (1..arr.len()).rev() {
        let j = rng.gen_range(0..=i);
        arr.swap(i, j);
    }
    let half = arr.len() / 2;
    AblConferences { east: arr[..half].to_vec(), west: arr[half..].to_vec() }
}
